### PR TITLE
Chore: Setup Knex.js

### PR DESCRIPTION
**Pull request by saszaCodes**
#### What does this PR do?
- sets up Knex.js, migration and seeding
- adds CRUD operations template class
- adds models for each table in database
- defines most of the routes defined in a [template document](https://docs.google.com/document/d/1cZqCmaFP--tI8nkjEOGY1x-Z-Wdn1oEUOsxeWACGlac/edit) and adds services handling conntections
#### Description of Task to be completed?
Setup Knex.js
#### How should this be manually tested?
To test if Knex.js is added properly and if migrations and seeding are set up, first run `npm run migrate:rollback`, this will drop all the tables in development db. Then, run `npm run migrate:latest` to add new tables and `npm run seed` to populate them. You can check if tables are created and populated correctly using RDBMS of your choice.

To test if routes, services and models are set up propetly, run `npm run dev` script to start up the server and use Postman to test if each endpoint defined in `documentsRoute.js` and `usersRoute.js` behaves as expected. Server expects any data to be send in a JSON format and responds with data in the same format.
#### Any background context you want to provide?
I may have overinterpreted what I was supposed to do in this task. I was following the [template document](https://docs.google.com/document/d/1cZqCmaFP--tI8nkjEOGY1x-Z-Wdn1oEUOsxeWACGlac/edit) attached in the *Data should be persisted to a RDBMS storage* Trello card, but after adding all the models, services and routes I realized it was probably a template for the entire project, not just this one task. But what's done is done, I don't want to roll back so many commits that I have spent a lot of time on, so hopefully the way I implemented my solutions doesn't go against your plans for the next tasks.

Also, I implemented the tables in db a little bit differently - I moved `username` and `password` from `user_logins` to `user_profiles` table, as it seems to me to fall logically in the same category - general info about the user - whereas data stored in `user_logins` should, I think, contain only information about user logins, not their details.
#### Questions:
- Have I followed the barrel pattern correctly with all the new files?
- Are there any security concerns with storing all user details - username, email, password and fullname - in the one table?
- How should modules with classes be exported - should I instantiate a new object at the bottom of the file and export it (that's the way I did it in this PR), or export just the class and create new object in the files I import it to instead (this seems to violate the DRY rule, but maybe there are other reasons to do it this way)?

### CODE REVIEW

File: src/lib/knex/migrations/20210917093623_user_profiles.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+      table.string('email').notNullable().unique();
+      table.string('fullname');
+      table.string('username').notNullable().unique();
+      table.string('password').notNullable();
+      table.timestamps(true, true);
+      table.date('deleted_at');
+    });
+}
+
+export function down(knex) {
```
[the remark this answers is not in the document]
Comment (saszaCodes) [reply src/lib/knex/migrations/20210917093623_user_profiles.js:12]: JSDoc left out intentionally

File: src/lib/knex/migrations/20210917093646_user_logins.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
+export function up(knex) {
+  return knex.schema
+    .createTable('user_logins', (table) => {
+      table.increments('id');
+      table.integer('user_profile_id').notNullable().references('id').inTable('user_profiles');
+      table.date('last_login');
+    });
+}
+
+export function down(knex) {
```
[the remark this answers is not in the document]
Comment (saszaCodes) [reply src/lib/knex/migrations/20210917093646_user_logins.js:12]: JSDoc left out intentionally

File: src/lib/knex/migrations/20210917093623_user_profiles.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
+export function up(knex) {
```
[the remark this answers is not in the document]
Comment (saszaCodes) [reply src/lib/knex/migrations/20210917093623_user_profiles.js:1]: JSDoc left out intentionally

File: src/lib/knex/migrations/20210917093653_documents.js [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+      table.increments('id');
+      table.string('title').notNullable();
+      table.text('body');
+      table.integer('author_id').notNullable().references('id').inTable('user_profiles');
+      table.timestamps(true, true);
+      table.date('deleted_at');
+    });
+}
+
+export function down(knex) {
```
[the remark this answers is not in the document]
Comment (saszaCodes) [reply src/lib/knex/migrations/20210917093653_documents.js:13]: JSDoc left out intentionally

File: src/lib/knex/seeds/01_user_profiles.js [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    username: faker.internet.userName(),
+    password: faker.internet.password(),
+  };
+  if (Math.random() < 0.15) {
+    data.deleted_at = faker.date.past();
+  }
+  fakeData.push(data);
+}
+
+export function seed(knex) {
```
[the remark this answers is not in the document]
Comment (saszaCodes) [reply src/lib/knex/seeds/01_user_profiles.js:15]: JSDoc left out intentionally

File: src/lib/knex/seeds/02_user_logins.js [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+const fakeData = [];
+const numberOfEntries = 200;
+for (let i = 0; i < numberOfEntries; i += 1) {
+  const data = {
+    last_login: faker.datatype.datetime(),
+  };
+  fakeData.push(data);
+}
+
+export function seed(knex) {
```
[the remark this answers is not in the document]
Comment (saszaCodes) [reply src/lib/knex/seeds/02_user_logins.js:14]: JSDoc left out intentionally

File: src/lib/knex/migrations/20210917093653_documents.js [hunk 7]
```diff
@@ -0,0 +1,16 @@
+export function up(knex) {
```
[the remark this answers is not in the document]
Comment (saszaCodes) [reply src/lib/knex/migrations/20210917093653_documents.js:1]: JSDoc left out intentionally

File: src/lib/knex/migrations/20210917093646_user_logins.js [hunk 8]
```diff
@@ -0,0 +1,13 @@
+export function up(knex) {
```
[the remark this answers is not in the document]
Comment (saszaCodes) [reply src/lib/knex/migrations/20210917093646_user_logins.js:1]: JSDoc left out intentionally

File: src/models/CRUD.js [hunk 9]
```diff
@@ -0,0 +1,95 @@
+import { knex as db } from '../lib';
+
+/** CRUD class contains template CRUD methods */
+export default class CRUD {
```
Comment (Avuidrauxs) [inline src/models/CRUD.js:5]: Here is the idea of how to implement an abstract class

```javascript
abstract class CrudOperations {
  public abstract create();

  public abstract update();

  public abstract delete();

  public abstract retrieve();
}

class Users extends CrudOperations {
  public create() {
    return null
  }
  
  public update() {
    return null
  }
  
  public delete() {
    return null
  }
  
  public retrieve() {
    return null
  }
}
```

File: src/models/documents.js [hunk 10]
```diff
@@ -0,0 +1,61 @@
+import CRUD from './CRUD';
+
+/** contains methods for CRUD operations in documents table */
+class Documents extends CRUD {
```
Comment (Avuidrauxs) [inline src/models/documents.js:5]: Will be nice to have a constructor where we initialize an instance of knex

```javascript
constructor(db) {
this.db = db; // Knex instance
}

async getAllDocuments () {

const documents = await this.db.select().from('documents').timeout(1000);
return documents
}
```

File: src/services/documentsService.js [hunk 11]
```diff
@@ -0,0 +1,81 @@
+import { userProfiles, documents } from '../models';
+
+export const documentsService = {
```
Comment (Avuidrauxs) [inline src/services/documentsService.js]: ### Nitpick

- Make your services classes also

File: src/services/usersService.js [hunk 12]
```diff
@@ -0,0 +1,163 @@
+import { userProfiles, userLogins } from '../models';
+
+export const usersService = {
```
Comment (Avuidrauxs) [inline src/services/usersService.js]: ### Nitpick

- Make it a class

**Review by Avuidrauxs — CHANGES_REQUESTED** [review]
Awesome job with the migration and seeding implementation

Just a few requested changes on how I want our classes implemented.

**Comment by saszaCodes** [comment]
@Avuidrauxs ready for review. I only didn't create a constructor where knex is initialized, as this led to unexpected behavior, which I'd like to discuss today

**Review by Avuidrauxs — APPROVED** [review]
Looks great!

**Pull request merged**